csharp/src/Apache.Arrow/DecimalUtility.cs

```diff
@@ -0,0 +1,163 @@
+            int drop = scale;
+            while (drop > PowersOfTenLength)
+            {
+                result /= s_powersOfTen[PowersOfTenLength];
+                drop -= PowersOfTenLength;
+            }
+
+            result /= s_powersOfTen[drop];
+            return result;
+        }
+
+        public static void GetBytes(BigInteger integerValue, int byteWidth, ref Span<byte> bytes)
+        {
+            if (bytes.Length != byteWidth)
+            {
+                throw new OverflowException("ValueBuffer size not equal to " + byteWidth + " byte width: " + bytes.Length);
+            }
+
+            Span<byte> integerBytes = integerValue.ToByteArray().AsSpan();
+            if (integerBytes.Length > byteWidth)
+            {
+                throw new OverflowException("Decimal size greater than " + byteWidth + " bytes: " + integerBytes.Length);
+            }
+
+            if (integerBytes.Length == byteWidth)
+            {
+                bytes = integerBytes;
+                return;
+            }
+
+            if (integerValue.Sign == -1)
+            {
+                integerBytes.CopyTo(bytes);
+                for (int i = integerBytes.Length; i < byteWidth; i++)
+                {
+                    bytes[i] = 255;
+                }
+            }
+            else
+            {
+                integerBytes.CopyTo(bytes);
+            }
+        }
+
+        public static bool CheckPrecisionAndScale(decimal value, int precision, int scale, out BigInteger integerValue)
+        {
+            DecimalLayout layout = new DecimalLayout(value); // use in place of decimal.GetBits(value) to avoid an allocation
+            integerValue = new BigInteger(BitConverter.GetBytes(layout.Lo).Concat(BitConverter.GetBytes(layout.Mid)).Concat(BitConverter.GetBytes(layout.Hi)).ToArray());
+
+            if (layout.Scale > scale)
+                throw new OverflowException("Decimal scale can not be greater than that in the Arrow vector: " + layout.Scale + " != " + scale);
+
+            if(integerValue >= BigInteger.Pow(10, precision))
+                throw new OverflowException("Decimal precision can not be greater than that in the Arrow vector: " + value + " has precision > " + precision);
+
+            if (layout.Scale < scale) // pad with trailing zeros
+            {
+                integerValue *= BigInteger.Pow(10, scale - layout.Scale);
+            }
+
+            if (value < 0) // sign the big int
+                integerValue = -integerValue;
+
+            return true;
+        }
+
+        [StructLayout(LayoutKind.Explicit)]
+        private readonly struct DecimalLayout
+        {
+            public DecimalLayout(decimal value)
+            {
+                this = default;
+                d = value;
+            }
+
+            [FieldOffset(0)] private readonly decimal d;
+
+            [FieldOffset(0)] private readonly int flags;
+            [FieldOffset(4)] private readonly int hi;
+#if BIGENDIAN
```

Comment:
This isn't going to work well because no one builds this assembly with `BIGENDIAN` enabled.

How about using:

```C#
        private readonly int _flags;
        private readonly uint _hi32;
        private readonly ulong _lo64;
```

Which is what System.Decimal uses?

To get the Low and the Mid, you can do:

```C#
internal uint Low => (uint)_lo64;
internal uint Mid => (uint)(_lo64 >> 32);
```

Reply:
I've removed this as part of the latest refactor